### PR TITLE
Add missing functions to new C++ API

### PR DESCRIPTION
**Pull request by ajreynol**
These functions are required for further work on parser migration.

I have yet to address most of the additions that will be needed for datatypes.

### CODE REVIEW

**Comment by ajreynol** [comment]
Will need several initial updates to datatypes before the unit tests for these functions can be written.

**Comment by ajreynol** [comment]
I've added unit tests for all functions.  One test for records currently causes segfaults, I've opened https://github.com/CVC4/cvc4-projects/issues/112 to track this.

File: test/unit/api/term_black.h [hunk 1]
```diff
@@ -652,3 +656,79 @@ void TermBlack::testTermAssignment()
...
+void TermBlack::testIsConst()
+{
+  Term x = d_solver.mkConst(d_solver.getIntegerSort(), "x");
+  Term one = d_solver.mkReal(1);
+  Term xpone = d_solver.mkTerm(PLUS, x, one);
+  Term onepone = d_solver.mkTerm(PLUS, one, one);
+  TS_ASSERT(!x.isConst());
+  TS_ASSERT(one.isConst());
+  TS_ASSERT(!xpone.isConst());
+  TS_ASSERT(!onepone.isConst());
```
Comment (aniemetz) [inline test/unit/api/term_black.h]: missing: `TS_ASSERT_THROWS` when Term is null

File: src/api/cvc4cpp.cpp [hunk 2]
```diff
@@ -1322,6 +1362,42 @@ bool Term::operator==(const Term& t) const { return *d_expr == *t.d_expr; }
...
+bool Term::operator<(const Term& t) const { return *d_expr < *t.d_expr; }
+
+bool Term::operator>(const Term& t) const { return *d_expr > *t.d_expr; }
+
+bool Term::operator<=(const Term& t) const { return *d_expr <= *t.d_expr; }
+
+bool Term::operator>=(const Term& t) const { return *d_expr >= *t.d_expr; }
+
+size_t Term::getNumChildren() const
+{
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/api/cvc4cpp.cpp]: Will do.

File: src/api/cvc4cpp.cpp [hunk 3]
```diff
@@ -1854,13 +1961,37 @@ DatatypeConstructor::~DatatypeConstructor() {}
...
+}
+
+DatatypeSelector DatatypeConstructor::operator[](size_t index) const
+{
+  return (*d_ctor)[index];
+}
+
 DatatypeSelector DatatypeConstructor::operator[](const std::string& name) const
 {
   // CHECK: selector with name exists?
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/api/cvc4cpp.cpp]: I think we need this check, I'll look into adding.

File: src/api/cvc4cpp.cpp [hunk 4]
```diff
@@ -2019,12 +2150,30 @@ Term Datatype::getConstructorTerm(const std::string& name) const
...
 bool Datatype::isParametric() const { return d_dtype->isParametric(); }
+bool Datatype::isCodatatype() const { return d_dtype->isCodatatype(); }
+
+bool Datatype::isTuple() const { return d_dtype->isTuple(); }
+
+bool Datatype::isRecord() const { return d_dtype->isRecord(); }
+
+bool Datatype::isFinite() const
+{
+  // CHECK: is resolved?
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/api/cvc4cpp.cpp]: Not sure, was just following comments from other functions.  It's not yet clear to me if the resolved checks are needed and if so how we should implement this.  I'll look into it.

File: src/api/cvc4cpp.cpp [hunk 5]
```diff
@@ -1816,39 +1931,68 @@ DatatypeConstructor::DatatypeConstructor() { d_ctor = nullptr; }
...
+{
+  return (*d_ctor)[index];
+}
+
 DatatypeSelector DatatypeConstructor::operator[](const std::string& name) const
 {
-  // CHECK: selector with name exists?
-  // CHECK: is resolved?
-  return (*d_ctor)[name];
+  return getSelector(name);
```
Comment (aniemetz) [inline src/api/cvc4cpp.cpp]: We want to introduce API tracing in the near future, and thus no API function should call an API function. A solution to this problem would be to move the functionality of `getSelector` into a private helper, and then both call the helper. That's also why the previous implementation did not call `getSelector` and that's why we have all the other helper functions.

File: src/api/cvc4cpp.cpp [hunk 6]
```diff
@@ -1960,39 +2105,55 @@ Datatype::~Datatype() {}
...
   CVC4_API_CHECK(idx < getNumConstructors()) << "Index out of bounds.";
   return (*d_dtype)[idx];
 }
 
 DatatypeConstructor Datatype::operator[](const std::string& name) const
 {
-  // CHECK: cons with name exists?
-  // CHECK: is resolved?
-  return (*d_dtype)[name];
+  return getConstructor(name);
```
Comment (aniemetz) [inline src/api/cvc4cpp.cpp]: Ditto

**Pull request merged**